rust/frameworks/tw_substrate/src/lib.rs

```diff
@@ -0,0 +1,82 @@
+use tw_coin_entry::error::prelude::*;
+use tw_ss58_address::NetworkId;
+
+pub mod entry;
+pub use entry::*;
+
+pub mod substrate_coin_entry;
+pub use substrate_coin_entry::*;
+
+pub mod address;
+pub use address::*;
+
+pub mod extrinsic;
+pub use extrinsic::*;
+
+pub mod extensions;
+pub use extensions::*;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum EncodeError {
+    InvalidNetworkId,
+    MissingCallIndices,
+    InvalidCallIndex,
+    InvalidAddress,
+    InvalidValue,
+    NotSupported,
+}
+
+impl EncodeError {
+    /// Returns a TWError with the given error and context.
+    pub fn with_context(self, context: String) -> TWError<Self> {
+        TWError::new(self).context(context)
+    }
+
+    /// Returns a TWResult with the given error and context.
+    pub fn tw_result<T>(self, context: String) -> TWResult<T, Self> {
+        Err(self.with_context(context))
+    }
+
+    /// Convert another error type to this error type and return a TWError.
+    pub fn with_error<T: std::fmt::Debug>(self, err: T) -> TWError<Self> {
+        TWError::new(self).context(format!("{err:?}"))
+    }
+}
```

Comment:
I tried to use those.  Some wouldn't keep the context or wouldn't use the previous error as context for the `EncodeError`.

Reply:
It was a little tricky to convert `EncodeError` to `TWResult` with a context without that `tw_result` method, but I was able to do it.  The two methods were only used once, so I removed them.  Now all 3 methods have been removed.

Reply:
Sorry for a long reply. Basically, you could call `Err(EncodeError).into_tw().context("...")`, because `into_tw()` function wraps `Result<T, Err1>` into `Result<T, TWError<Err2>>`, where Err2: From<Err2>